File: DataAccessLayer/src/test/java/com/zeyad/genericusecase/realm_test_models/TestModel.java
package com.zeyad.genericusecase.realm_test_models;

import android.support.annotation.NonNull;

import com.google.gson.annotations.SerializedName;

import io.realm.RealmObject;
import io.realm.annotations.PrimaryKey;

public class TestModel extends RealmObject {

    @SerializedName("id")
    @PrimaryKey
    private int id;
    @SerializedName("value")
    private String value;

    public TestModel(int id, String value) {
        this.id = id;
        this.value = value;
        if (id <= 0)
            throw new IllegalArgumentException("id should be greater than 0");
    }

    public TestModel() {
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getValue() {
        return value;
    }

    public void setValue(String value) {
        this.value = value;
    }

    @NonNull
    @Override
    public String toString() {
        return "TestModel{" +
                "id=" + id +
                ", value='" + value + '\'' +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof TestModel)) return false;
        TestModel testModel = (TestModel) o;
        return getId() == testModel.getId() && getValue().equals(testModel.getValue());
    }

    @Override
    public int hashCode() {
        int result = getId();
        result = 31 * result + getValue().hashCode();
        return result;
    }
}

File: usecases/src/test/java/com/zeyad/usecases/TestModel.kt
package com.zeyad.usecases

import android.arch.persistence.room.Entity
import android.arch.persistence.room.PrimaryKey
import com.google.gson.annotations.SerializedName

/**
 * @author by ZIaDo on 2/13/17.
 */
@Entity
class TestModel {
    @SerializedName("id")
    @PrimaryKey
    var id: Int = 0

    @SerializedName("value")
    var value: String? = null

    constructor(id: Int, value: String) {
        this.id = id
        this.value = value
        if (id <= 0) throw IllegalArgumentException("id should be greater than 0")
    }

    constructor()

    override fun toString(): String {
        return "TestModel{" + "id=" + id + ", value='" + value + '\''.toString() + '}'.toString()
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o !is TestModel) return false
        val testRealmObject = o as TestModel?
        return id == testRealmObject!!.id && value == testRealmObject.value
    }

    override fun hashCode(): Int {
        var result = id
        result = 31 * result + value!!.hashCode()
        return result
    }
}
